core/errors.go

```diff
@@ -0,0 +1,206 @@
+  *IBMError
+  OperationID string
+  Response *DetailedResponse
+  ErrorCode string
+}
+
+func (e *HTTPError) GetDebugMessage() string {
+  return getErrorInfoAsYAML(e)
+}
+func (e *HTTPError) GetID() string {
+  return createIDHash("http_error", e.getBaseSignature(), e.OperationID, fmt.Sprint(e.Response.GetStatusCode()), e.ErrorCode)
+}
+
+// Error creation functions
+
+func ibmErrorf(err error, summary, system, version, discriminator string) *IBMError {
+  if summary == "" {
+    summary = err.Error()
+  }
+  return &IBMError{
+    Summary: summary,
+    System: system,
+    Version: version,
+    discriminator: discriminator,
+    causedBy: err,
+  }
+}
+
+// Convenience function to create local instances of `SDKError`
+func coreSDKErrorf(err error, summary, discriminator, function string) *SDKError {
+  system, version := getSystemInfo()
+  return &SDKError{
+    IBMError: ibmErrorf(err, summary, system, version, discriminator),
+    Function: function,
+  }
+}
+
+// Create new instances of `SDKError`
+func SDKErrorf(err error, summary, system, version, discriminator, function string) *SDKError {
+  return &SDKError{
+    IBMError: ibmErrorf(err, summary, system, version, discriminator),
+    Function: function,
+  }
+}
+
+// Convenience function to create local instances `HTTPError`
+func coreHTTPErrorf(err error, summary, discriminator, operationID, code string, response *DetailedResponse) *HTTPError {
+  system, version := getSystemInfo()
+  return &HTTPError{
+    IBMError: ibmErrorf(err, summary, system, version, discriminator),
+    OperationID: operationID,
+    Response: response,
+    ErrorCode: code,
+  }
+}
+
+// Create new instances of `HTTPError`
+func HTTPErrorf(err error, summary, system, version, discriminator, operationID, code string, response *DetailedResponse) *HTTPError {
+  return &HTTPError{
+    IBMError: ibmErrorf(err, summary, system, version, discriminator),
+    OperationID: operationID,
+    Response: response,
+    ErrorCode: code,
+  }
+}
+
+func rewrapSDKError(err error, function string) error {
+  // Keep error details but note the function it is returning from.
+  if err != nil {
+    if sdkError, ok := err.(*SDKError); ok {
+      sdkError.Function = function
+      err = sdkError
+    }
+  }
+
+  return err
+}
+
+// TODO: consider accepting options model for future-oriented flexibility.
+// ^ The above can apply to all functions in this file.
+func RewrapClientError(err error, system, version, function, operationID string) error {
+    if err == nil {
+      return err
+    }
+
+    // Keep error details but note the new context
+    if sdkError, ok := err.(*SDKError); ok {
+      sdkError.System = system
+      sdkError.Version = version
+      sdkError.Function = function
+      err = sdkError
+    } else if httpError, ok := err.(*HTTPError); ok {
+      httpError.System = system
+      httpError.Version = version
+      httpError.OperationID = operationID
+      err = httpError
+    }
```

Comment:
You can use type switch here which would improve the readability a little - in my opinion. Maybe it's just my personal preference, I let you to decide. :)
```suggestion
	switch _err := err.(type) {
	case SDKError:
		_err.System = system
		_err.Version = version
		_err.Function = function
		err = _err
	case HTTPError:
		_err.System = system
		_err.Version = version
		_err.OperationID = operationID
		err = _err
	}
```

Reply:
That's a good call, I like it! I'll make that change